locator/tablets.cc

```diff
@@ -179,4 +181,129 @@ size_t tablet_metadata::external_memory_usage() const {
+        auto&& tablets = get_tablet_map();
+        auto tablet = tablets.get_tablet_id(search_token);
+        auto&& replicas = tablets.get_tablet_info(tablet).replicas;
+        tablet_logger.trace("get_natural_endpoints({}): table={}, tablet={}, replicas={}", search_token, _table, tablet, replicas);
+        return to_replica_set(replicas);
+    }
+
+    virtual inet_address_vector_replica_set get_natural_endpoints_without_node_being_replaced(const token& search_token) const override {
+        auto result = get_natural_endpoints(search_token);
+        maybe_remove_node_being_replaced(*_tmptr, *_rs, result);
+        return result;
+    }
+
+    virtual inet_address_vector_topology_change get_pending_endpoints(const token& search_token, const sstring& ks_name) const override {
+        auto&& tablets = get_tablet_map();
+        auto tablet = tablets.get_tablet_id(search_token);
+        auto&& info = tablets.get_tablet_transition_info(tablet);
+        if (!info) {
+            return {};
+        }
+        tablet_logger.trace("get_pending_endpoints({}): table={}, tablet={}, replica={}",
+                            search_token, _table, tablet, info->pending_replica);
+        return {get_endpoint_for_host_id(info->pending_replica.host)};
+    }
+
+    virtual std::unique_ptr<token_range_splitter> make_splitter() const override {
+        class splitter : public token_range_splitter {
+            token_metadata_ptr _tmptr; // To keep the tablet map alive.
+            const tablet_map& _tmap;
+            tablet_id _next;
+        public:
+            splitter(token_metadata_ptr tmptr, const tablet_map& tmap)
+                : _tmptr(std::move(tmptr))
+                , _tmap(tmap)
+            { }
+
+            void reset(dht::ring_position_view pos) override {
+                _next = _tmap.get_tablet_id(pos.token());
+            }
+
+            std::optional<dht::token> next_token() override {
+                if (_next == _tmap.tablet_count()) {
+                    return std::nullopt;
+                }
+                return _tmap.get_last_token(_next++);
+            }
+        };
+        return std::make_unique<splitter>(_tmptr, get_tablet_map());
+    }
+};
+
+size_t tablet_aware_replication_strategy::parse_initial_tablets(const sstring& val) const {
+    try {
+        return std::stol(val);
+    } catch (...) {
+        throw exceptions::configuration_exception(format("\"initial_tablets\" must be numeric; found {}", val));
+    }
+}
+
+void tablet_aware_replication_strategy::validate_tablet_options(const gms::feature_service& fs,
+                                                                const replication_strategy_config_options& opts) const {
+    for (auto& c: opts) {
+        if (c.first == "initial_tablets") {
+            if (!fs.tablets) {
+                throw exceptions::configuration_exception("Tablet replication is not enabled");
+            }
+            parse_initial_tablets(c.second);
```

Comment:
It's there so that abstract_replication_strategy::validate_options() validates all the options. The question is more general actually, why do we need validate_options() if we parse the options in the constructor anyway, and it applies not only to tablet options, but other options as well. I don't have an answer for that, I just played along.